doc/random.adoc

```diff
@@ -0,0 +1,193 @@
+
+Minetest provides four different random sources, each with its own merits.
+Modders must choose wisely unless they can let the engine do the random for them
+(e.g. randomly picking a sound or a texture for particles).
+
+== Lua builtins
+
+Not restricted by mod security, these functions are available to both SSMs and CSMs:
+
+=== https://www.lua.org/manual/5.1/manual.html#pdf-math.randomseed[`math.randomseed`]
+
+Seed the random. Minetest already does this for you using the system time.
+
+[IMPORTANT]
+.Do not seed the random
+====
+Do not seed the random to turn it into a deterministic random source
+as other mods may expect it to be "non-deterministic".
+
+Conversely, do not rely on the random to have any particular seed either; other mods & the engine
+may have seeded it (using the system time) to be "non-deterministic".
+
+The problem with `math.randomseed` is that there is only one global, hidden seed.
+There is no way to get the current seed out; mods can't restore their random sequence.
+Mods seeding the random thus necessarily conflict - unless they all expect
+it to be "non-deterministic" and only seed it accordingly
+(ideally not at all, since the engine-side seeding should suffice).
+
+If you need `math.random` for its performance but want it to be deterministic,
+you may *reseed* the random after you're done with it to ensure that it is "non-deterministic" again.
+
+[source,lua]
+---
+-- Use the random to generate a seed for the random; preferable over using system time,
+-- as the latter may be deterministic
+local seed = ... -- some fixed seed
+local reseed = math.random(2^31-1)
+math.randomseed(seed) -- temporarily make the random "deterministic"
+-- ... do something using `math.random` ...
+math.randomseed(reseed)
+---
+====
+
+=== https://www.lua.org/manual/5.1/manual.html#pdf-math.random[`math.random`]
+
+Get a random number. Very versatile; allows getting floats between `0` and `1` or integers in a range.
+
+NOTE: The random numbers between `0` and `1` do not provide a full 52-bit mantissa
+full of entropy; they usually have around 32 bits of entropy.
+
+WARNING: When using this to obtain integers,
+make sure that both the upper & lower bound
+as well as their difference are within the C `int` range -
+otherwise you may get overflows & errors.
+
+TIP: Use `math.random` as your go-to versatile "non-deterministic" random source.
+
+== Random Number Generators
+
+=== `PcgRandom`
+
+A seedable 32-bit signed integer pseudo-random number generator.
+
+==== `PcgRandom(seed)`
+
+Constructs a `PcgRandom` instance with the given seed,
+which should be an integer within 32-bit bounds.
+
+==== Methods
+
+===== `:next([min, max])`
+
+If `min` and `max` are both omitted,
+they default to `-2^31` (`-2147483648`)
+and `2^31 - 1` (`2147483647`) respectively.
+
+===== `:rand_normal_dist(min, max, [num_trials])`
+
+WARNING: No successful use of this function is documented. Consider implementing your own normal distribution instead.
```

Comment:
Usage in mods: https://web.archive.org/web/20231224131713if_/https://content.minetest.net/zipgrep/1b9f9e78-99da-4c02-9f75-f6134f362fa7/